internal/pkg/text/template/parse/node.go

```diff
@@ -195,25 +195,34 @@ func (p *PipeNode) append(command *CommandNode) {
 
 func (p *PipeNode) String() string {
 	var sb strings.Builder
-	p.writeTo(&sb, false)
+	p.writeTo(&sb, false, &j2FuncContext{})
 	return sb.String()
 }
 
-func (p *PipeNode) writeTo(sb *strings.Builder, isConditional bool) {
+type j2FuncContext struct {
+	IsFunc        bool
+	PipeNodeCount int
+}
+
+func (p *PipeNode) writeTo(sb *strings.Builder, isConditional bool, funcContext *j2FuncContext) {
 	if len(p.Decl) > 0 {
 		for i, v := range p.Decl {
 			if i > 0 {
 				sb.WriteString(", ")
 			}
-			v.writeTo(sb, isConditional)
+			v.writeTo(sb, isConditional, funcContext)
 		}
 		sb.WriteString(" := ")
 	}
 	for i, c := range p.Cmds {
 		if i > 0 {
 			sb.WriteString(" | ")
 		}
-		c.writeTo(sb, isConditional)
+		injectedFuncContext := j2FuncContext{
```

Comment:
paramater  funcContext *j2FuncContext is required, initialized, and used only form pipeNode write method, You should introduced a new method for this since other objects don't have to fulfill this contract. It is more of a design creep if it's not having any meaning to other methods. I think the same goes for isconditional .

Reply:
Respectfully, I disagree.  This is not design creep;  it is required.  Passing context information through a struct is a common tactic in Abstract Syntax Tree design;  the fact that the original author had a different notion of what the API should look like does not and should not prevent us from modifying as we see fit, especially in our forked non-exported implementation.  

The reason that this is made difficult is because several Node types are overloaded in the text/template Abstract Syntax Tree;  for example, PipeNode is used to represent conditionals as well as sprig function invocations.  To properly handle these orthogonal cases, we should have not forked text/template, and rather built our own AST which aggregates separate Node Types for the various use cases, each of which has the knowledge to properly output the given node types (i.e., ConditionalPipeNode and TemplateFunctionPipeNode).  However, in the interest of time, that is not what we decided.  We would have likely had to rewrite the lexer too in that case, since the lexer is responsible for defining the grammar to translate input symbols into corresponding Nodes, but that is simply infeasible given our timing constraints.

Why is the method signature change required?  Because separating out a unique "writeTo" method would not be limited to just the PipeNode type.  In fact, we would have to fork any possible candidate which requires this context.  For example, PipeNode invokes CommandNode.writeTo(...) for conditionals and template functions.  CommandNode then calls "writeTo" on several different types, including:
* IdentifierNode for the function call.
* FieldNode for .Values. variabls
* NumberNode for numbers
So each of those would require this additional context information, so you would need to fork each of them.  Then, once you throw in nesting of PipeNode Commands, you realize that the context must be persisted across piped invocations too, so you really do need to change the base Node.writeTo method signature.

My point is that this is the path of least resistance, and the only realistic way to achieve this goal outside of using a global variable (even worse in my opinion).  We could simplify this API in a follow up patch to include IsConditional as a field in j2FuncContext struct, so we are only adding a single parameter in total.  The point is, we would have had to add this context regardless, whether implied through abstracting our own NodeType(s) in the AST, or through passing additional context through the method calls.  I recommend pushing this in as is, and we can investigate ways to simplify the method signature at a later time.